Make DefaultProblemDetailsFactory public

GiovanniBraconi opened the pull request:
<!-- Thank you for submitting a pull request to our repo. -->

<!-- If this is your first PR in the ASP.NET Core repo, please run through the checklist
below to ensure a smooth review and merge process for your PR. -->

- [x] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [ ] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

<!-- Once all that is done, you're ready to go. Open the PR with the content below. -->


## Description

 Fixes #53030

Following this [suggestion](https://github.com/dotnet/aspnetcore/issues/53030#issuecomment-1877918398) i changed the DefaultProblemDetailsFactory class access modifier from internal to public.

GiovanniBraconi committed:
Make DefaultProblemDetailsFactory public

GiovanniBraconi committed:
Add XML comments to DefaultProblemDetailsFactory.cs

gfoidl commented on src/Mvc/Mvc.Core/src/Infrastructure/DefaultProblemDetailsFactory.cs:
```diff
@@ -9,20 +9,37 @@
+/// It provides methods to create instances of `ProblemDetails` and `ValidationProblemDetails` with default settings.
+/// This class uses the provided `ApiBehaviorOptions` for client error mapping and an optional custom configuration action to further customize the problem details.
+/// </summary>
+public sealed class DefaultProblemDetailsFactory : ProblemDetailsFactory
```
@GiovanniBraconi please address the analyzer-failure.
Have a look at [How to use Microsoft.CodeAnalysis.PublicApiAnalyzers](https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md) and see at the files in https://github.com/dotnet/aspnetcore/tree/main/src/Mvc/Mvc.Core/src folder.

GiovanniBraconi committed:
Fix analyzer errors

- Add DefaultProblemDetailsFactory.cs constructor and methods to PublicAPI.Unshipped.txt

- Remove duplicated xml comments from methods

captainsafia commented on src/Mvc/Mvc.Core/src/Infrastructure/DefaultProblemDetailsFactory.cs:
```diff
@@ -95,6 +108,6 @@ private void ApplyProblemDetailsDefaults(HttpContext httpContext, ProblemDetails
         }
 
-        _configure?.Invoke(new() { HttpContext = httpContext!, ProblemDetails = problemDetails });
+        _configure?.Invoke(new() {HttpContext = httpContext!, ProblemDetails = problemDetails});
```
Nit: our style guidelines recommend spacing between braces here.

```suggestion
        _configure?.Invoke(new() { HttpContext = httpContext!, ProblemDetails = problemDetails });
```

captainsafia reviewed (approved):
LGTM outside of a small nit to apply.

We already have test coverage for this via [ProblemDetailsFactoryTest](https://source.dot.net/#Microsoft.AspNetCore.Mvc.Core.Test/Infrastructure/ProblemDetailsFactoryTest.cs,e16b710dae9e3e36).

Thanks for the first look @gfoidl!

GiovanniBraconi committed:
Add spacing between braces

Co-authored-by: Safia Abdalla <safia@safia.rocks>

uhfath commented:
Is there any chance of rolling this back to 8.x sometime in the future?

martincostello commented:
New features aren't backported, so highly unlikely.

uhfath commented:
It's a pity to hear that.
